<a href="https://colab.research.google.com/github/crazygovind/AI-Foundation-Whitehat-Jr/blob/master/17_Project_Pandas_DataFrame_Applied_Tech_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Instructions

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concepts covered in the following lessons:

1. While Loop, Data-Type Conversion and Conditional Statements.

2. Pandas DataFrame.


---

### Activities


  

#### Activity 1: Create Pandas DataFrame

Create a pandas DataFrame by using the below link which has the dataset of NBA players in the CSV format: 

**Dataset** : https://media.geeksforgeeks.org/wp-content/uploads/nba.csv

Also, print the first five rows of the dataset.

In [ ]:
# Create a pandas DataFrame
import pandas as pd
df=pd.read_csv('https://media.geeksforgeeks.org/wp-content/uploads/nba.csv')

---

#### Activity 1.2 : Check for Null values

Check for missing values in the DataFrame.

In [ ]:
# Check for missing values in the above Pandas series.

print(df.isnull())
df.isnull().sum()

      Name   Team  Number  Position    Age  Height  Weight  College  Salary
0    False  False   False     False  False   False   False    False   False
1    False  False   False     False  False   False   False    False   False
2    False  False   False     False  False   False   False    False    True
3    False  False   False     False  False   False   False    False   False
4    False  False   False     False  False   False   False     True   False
..     ...    ...     ...       ...    ...     ...     ...      ...     ...
453  False  False   False     False  False   False   False    False   False
454  False  False   False     False  False   False   False     True   False
455  False  False   False     False  False   False   False     True   False
456  False  False   False     False  False   False   False    False   False
457   True   True    True      True   True    True    True     True    True

[458 rows x 9 columns]


Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

**Q:** Are there any missing values in the DataFrame ?

**A:**  yes

---

#### Activity 1.3 : Get the Number of rows and columns

Get the number of rows and columns of the DataFrame.

In [ ]:
# Get the Number of rows and columns
df.shape

(458, 9)

**Q:** How many rows are there in the DataFrame?

**A:** 458 rows and 9 columns


---

#### Activity 1.4: Create Pandas series for the player 'Avery Bradley'

Using the `iloc[]` function, create a Pandas series for the player Avery Bradley and store it in a variable called `avery_bradley`.

In [ ]:
# Create pandas series for the player 'Avery Bradley'

df['Name']=='Avery Bradley'

0       True
1      False
2      False
3      False
4      False
       ...  
453    False
454    False
455    False
456    False
457    False
Name: Name, Length: 458, dtype: bool

**Q**: Which team does Avery Bradely plays for?

**A**: 

---

#### Activity 2: Find the number of solutions for a Quadratic equation

A quadratic equation $ax² + b x + c = 0$ has either 0, 1, or 2 distinct solutions for real values of $x$. Given $a$, $b$ and $c$, you should only return the number of solutions that we can get from the equation.

The above quadratic equation is solved using the formula, \begin{array}{*{20}c} {x = \frac{{ - b \pm \sqrt {b^2 - 4ac} }}{{2a}}} & {{\rm{when}}} & {ax^2 + bx + c = 0} \\ \end{array}

The `±` in the above formula means there are **TWO** answers.

Do you see $b^2 − 4ac$ in the formula above? It is called the **Discriminant**, because it can "discriminate" between the possible types of answer:

  - when $b^2 − 4ac$ is **positive**, we get **two** Real solutions.

  - when $b^2 − 4ac$ is **zero**, we get just **one** real solution. 

  - when $b^2 − 4ac$ is **negative**, we get a pair of **complex solution**.


Examples:
```
1. solutions(1, 0, -1) ➞ 2
# x² - 1 = 0 has two solutions (x = 1 and x = -1).

2. solutions(1, 0, 1) ➞ 0
# x² + 1 = 0 has no solutions.
```

**Note**: You do not need to calculate the solutions, just return how many solutions are there by calculating the value of $b^2 − 4ac$.


In [1]:
# Number of solutions for a quadratic equation
discriminant=(1,5,-2)

if ((discriminant[1])**2 - 4*(discriminant[0])*(discriminant[2])) > 0:
  print('Equation has two solutions')
elif ((discriminant[1])**2 - 4*(discriminant[0])*(discriminant[2])) == 0:
  print('Equation has one real solution')
elif ((discriminant[1])**2 - 4*(discriminant[0])*(discriminant[2])) < 0:
  print('Equation has no solution')

Equation has two solutions


---